documentation/modules/ROOT/pages/connectors/cassandra.adoc

```diff
@@ -167,14 +167,12 @@ For example, consider a Cassandra installation with an `inventory` keyspace that
 [[cassandra-schema-evolution]]
 === Schema Evolution
 
-DDLs are not recorded in commit logs. When the schema of a table change, this change is issued from one of the Cassandra node and propagated to other nodes via Gossip Protocol. This implies detection of schema changes are achieved on a best-effort basis. This is done by periodically polling the schema of each cdc-enabled table in the cluster via a Cassandra driver, and then update the cached version of the schema. Because of this implementation, if a new column is added to a table and then writes are issued against that column immediately, it is possible that data from that column will not be reflected in the CDC event. This is why it is recommened to pause for some time (configured with `schema.refresh.interval.ms`) after issuing a schema change.
+DDLs are not recorded in commit logs. When the schema of a table change, this change is issued from one of the Cassandra node and propagated to other nodes via Gossip Protocol.
 
-**TODO**: it may be possible to reactively refresh schema whenever an unexpect column appears in a mutation to improve schema change detection; worth looking into.
+Schema changes in Cassandra will be detected less than 1s by the implemented SchemaChangeListener, which will then update the Schema Instance loaded from Cassandra as well as the Kafka key value schemas cached upon starting.
```

Comment:
@bingqinzhou, here I think it's worth adding a sentence or two on the situation we were discussing yesterday: handling a change event that was created _before_ a schema change which already caused an update of the connector's internal schema metadata. IIUC, the connector would currently fail in that situation, so can we describe the procedure to get it unstuck in this case? I suppose one would have to somehow skip over those events of the table created before the schema change?

Reply:
Taking a step back, I'm still wondering whether there's a way to obtain the validity of a schema change (like `OffsetPosition`), so we can correlate change events to the right schema version. Ultimately, this may require changes in Cassandra itself? // CC @smiklosovic

Reply:
Ah, so reading your comment in https://issues.redhat.com/browse/DBZ-3417 just now; is this to say that the connector will not necessarily fail actually in the situation described above? I think the docs should state the different kinds of changes and the implications:

* CDC gets disabled for table: some changes from before disabling it may not end up being propagated
* Column gets removed: some change events from before removing the column may already miss that column although it still was present at that time
* Column gets added: ?
* Column type gets changed: ?

Reply:
If a column is added and this change is propagated, I would expect that in the body / payload of  that event there would be "null" value for a new column when we still read the remaining of commit logs which were created before that change happened.

If a column is removed, then this column will not be added into body / payload even though data are there.

I do not think that you can change datatype on vanilla Cassandra, it is possible to do that with Datastax distribution though.

Right now, the information about wether a mutation is cdc-enabled comes from schema which we synchronise and so on. This is the fundamental design flaw I would say because then we see that cdc is set to false, mutations which were sent in time it was still true are not processed.

The solution to this would be to serialise CDC flag directly to mutation so serialised mutation would carry this information in itself. This is possible to do but it would change the format of mutations / sstables, it would suddenly be a different version and you could not migrate between versions easilly. This might be maybe implemented in Cassandra 4 but it still places a lot of complexity on end users when it comes to different versions and so on, at least that is my take on this.

Reply:
> Ah, so reading your comment in https://issues.redhat.com/browse/DBZ-3417 just now; is this to say that the connector will not necessarily fail actually in the situation described above? I think the docs should state the different kinds of changes and the implications:
> 
> * CDC gets disabled for table: some changes from before disabling it may not end up being propagated
> * Column gets removed: some change events from before removing the column may already miss that column although it still was present at that time
> * Column gets added: ?
> * Column type gets changed: ?

- CDC gets disabled for table: Changes from before disabling it will be skipped. 
- Column gets removed: Attempts to deserialize changes involving the removed column before it's removed will fail with unrecoverable error, but Cassandra Connector will not stop, the error will be logged, recorded and skipped. 
- Column gets added: Work totally well.
- Column type gets changed: This kind of schema change is not allowed in Cassandra.

Reply:
> @bingqinzhou, here I think it's worth adding a sentence or two on the situation we were discussing yesterday: handling a change event that was created _before_ a schema change which already caused an update of the connector's internal schema metadata. IIUC, the connector would currently fail in that situation, so can we describe the procedure to get it unstuck in this case? I suppose one would have to somehow skip over those events of the table created before the schema change?

@gunnarmorling Updated the doc to include the situations which don't work well with the new approach. Cassandra Connector will not stop upon these situations, just skip the changes. Compared with severe data loss issues brought by the old approach, I think the effects of the situations here can be ignored.